cmd/cmd.go

```diff
@@ -70,3 +80,323 @@ func validateUserConfig(cfg *viper.Viper) error {
 	}
 	return nil
 }
+
+// sanitizeLegacyNumericSuffixFilepath is a workaround for a path bug due to an early design
+// decision (later reversed) to allow numeric suffixes for exercise directories,
+// allowing people to have multiple parallel versions of an exercise.
+func sanitizeLegacyNumericSuffixFilepath(file, slug string) string {
+	pattern := fmt.Sprintf(`\A.*[/\\]%s-\d*/`, slug)
+	rgxNumericSuffix := regexp.MustCompile(pattern)
+	if rgxNumericSuffix.MatchString(file) {
+		file = string(rgxNumericSuffix.ReplaceAll([]byte(file), []byte("")))
+	}
+	// Rewrite paths submitted with an older, buggy client where the Windows
+	// path is being treated as part of the filename.
+	file = strings.Replace(file, "\\", "/", -1)
+	return filepath.FromSlash(file)
+}
+
+// download is a download from the Exercism API.
+type download struct {
+	*downloadParams
+	*downloadPayload
+	*downloadWriter
+}
+
+// newDownloadFromExercise is a convenience wrapper for creating a new download.
+func newDownloadFromExercise(usrCfg *viper.Viper, exercise workspace.Exercise) (*download, error) {
+	downloadParams, err := newDownloadParamsFromExercise(usrCfg, exercise)
+	if err != nil {
+		return nil, err
+	}
+	return newDownload(downloadParams)
+}
+
+// newDownloadFromFlags is a convenience wrapper for creating a new download.
+func newDownloadFromFlags(usrCfg *viper.Viper, flags *pflag.FlagSet) (*download, error) {
+	downloadParams, err := newDownloadParamsFromFlags(usrCfg, flags)
+	if err != nil {
+		return nil, err
+	}
+	return newDownload(downloadParams)
+}
+
+// newDownload initiates a download by requesting a downloadPayload from the Exercism API.
+func newDownload(params *downloadParams) (*download, error) {
+	if err := params.validate(); err != nil {
+		return nil, err
+	}
+	d := &download{downloadParams: params}
+	d.downloadWriter = &downloadWriter{download: d}
```

Comment:
Having circular dependencies seems weird (`Download` and `downloadWriter` both hold direct references to each other) but maybe this is ok? It reads better to me to have the separate types but maybe doing so is too hacky. The thinking is that `Download` mainly contains data and queries whereas `downloadWriter` distinguishes commands (the write methods). This might offer flexibility as it could be easily modified to be an interface if needed (though maybe I'm missing YAGNI).